contracts/non-fungible-tokens/andromeda-crowdfund/src/testing/tests.rs

```diff
@@ -1,179 +1,141 @@
+    WasmMsg,
 };
 
 use crate::{
     contract::{execute, instantiate},
     state::{CAMPAIGN_CONFIG, TIERS, TIER_ORDERS},
-    testing::mock_querier::mock_dependencies_custom,
+    testing::mock_querier::{mock_dependencies_custom, mock_zero_price_tier, MOCK_DEFAULT_LIMIT},
 };
 
 use super::mock_querier::{mock_campaign_config, mock_campaign_tiers};
 
-fn init(deps: DepsMut, modules: Option<Vec<Module>>) -> Response {
-    let config = mock_campaign_config();
-    let tiers = mock_campaign_tiers();
+fn init(deps: DepsMut, config: CampaignConfig, tiers: Vec<Tier>) -> Response {
     let msg = InstantiateMsg {
         campaign_config: config,
         tiers,
         owner: None,
-        modules,
+        modules: None,
         kernel_address: MOCK_KERNEL_CONTRACT.to_string(),
     };
 
     let info = mock_info("owner", &[]);
     instantiate(deps, mock_env(), info, msg).unwrap()
 }
 
-#[test]
-fn test_instantiate() {
-    let mut deps = mock_dependencies_custom(&[]);
+fn get_tiers(storage: &dyn Storage) -> Vec<Tier> {
+    TIERS
+        .range_raw(storage, None, None, Order::Ascending)
+        .map(|res| res.unwrap().1)
+        .collect()
+}
+
+#[cfg(test)]
+mod test {
 
-    let res = init(deps.as_mut(), None);
+    use super::*;
 
-    assert_eq!(
+    fn instantiate_response(owner: &str) -> Response {
         Response::new()
             .add_attribute("method", "instantiate")
             .add_attribute("type", "crowdfund")
             .add_attribute("kernel_address", MOCK_KERNEL_CONTRACT)
-            .add_attribute("owner", "owner"),
-        res
-    );
-
-    assert_eq!(
-        mock_campaign_config(),
-        CAMPAIGN_CONFIG.load(deps.as_mut().storage).unwrap()
-    );
-
-    let tiers: Vec<Tier> = TIERS
-        .range_raw(deps.as_ref().storage, None, None, Order::Ascending)
-        .map(|res| res.unwrap().1)
-        .collect();
-    assert_eq!(mock_campaign_tiers(), tiers);
-}
-
-#[test]
-fn test_instantiate_invalid_tiers() {
-    let mut deps = mock_dependencies_custom(&[]);
-    let config = mock_campaign_config();
-    let mut tiers = mock_campaign_tiers();
-    tiers.push(Tier {
-        level: Uint64::new(1u64),
-        label: "Tier 1".to_string(),
-        limit: Some(Uint128::new(100)),
-        price: Uint128::zero(),
-        meta_data: TierMetaData {
-            extension: TokenExtension {
-                publisher: MOCK_ADO_PUBLISHER.to_string(),
+            .add_attribute("owner", owner)
+    }
+
+    struct InstantiateTestCase {
+        config: CampaignConfig,
+        tiers: Vec<Tier>,
+        expected_res: Result<Response, ContractError>,
+        expected_config: Option<CampaignConfig>,
+        expected_tiers: Vec<Tier>,
+    }
+    #[test]
+    fn test_instantiate() {
+        let test_cases: Vec<InstantiateTestCase> = vec![
+            InstantiateTestCase {
+                config: mock_campaign_config(),
```

Comment:
Anywhere we use test cases we should add a label, usually `name` to identify what the case is doing and provide it alongside any assertions so we know which case is broken.

Reply:
Resolved